metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java

```diff
@@ -786,12 +773,29 @@ public void run() throws Exception {
         }
     }
 
+    class RecoverMigrationStateFromZKEvent extends MigrationEvent {
+        @Override
+        public void run() throws Exception {
+            applyMigrationOperation("Recovering migration state from ZK", zkMigrationClient::getOrCreateMigrationRecoveryState);
+            String maybeDone = migrationLeadershipState.initialZkMigrationComplete() ? "done" : "not done";
+            log.info("Initial migration of ZK metadata is {}.", maybeDone);
+
+            // Once we've recovered the migration state from ZK, install this class as a metadata publisher
+            // by calling the initialZkLoadHandler.
+            initialZkLoadHandler.accept(KRaftMigrationDriver.this);
+
+            // Transition to INACTIVE state and wait for leadership events.
+            transitionTo(MigrationDriverState.INACTIVE);
+        }
+    }
+
     class PollEvent extends MigrationEvent {
+
         @Override
         public void run() throws Exception {
             switch (migrationState) {
                 case UNINITIALIZED:
-                    recoverMigrationStateFromZK();
+                    eventQueue.append(new RecoverMigrationStateFromZKEvent());
```

Comment:
Should we use `prepend` to make sure this event is executed ASAP

Reply:
On second thought, I don't think my question makes sense. The following `PollEvent` can only after `RecoverMigrationStateFromZKEvent` finishes.